Test greetz.nl website

HaykMelikyan opened the pull request:
-Add item to favorites and check it's appearance
-Change amount and check total price calculation

HaykMelikyan committed:
Test greetz.nl website
-Add item to favorites and check it's appearance
-Change amount and check total price calculation

chandranik90 commented on src/test/java/greetzWebsiteTest.java:
```diff
@@ -0,0 +1,121 @@
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class greetzWebsiteTest {
```
Class name should start by upper-case

HaykMelikyan committed:
Replace Selenium waits with Thread.sleep

HamTorosyan commented on src/test/java/greetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class greetzWebsiteTest {
```
Class name starts with Capital letter

HaykMelikyan replied:
fixed

HamTorosyan commented on src/test/java/greetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+
+        // Get first product's name and price
+        WebElement productItem = driver.findElement(By.cssSelector("div.b-products-grid__item:nth-child(1)"));
+        String productNameExpected = productItem.findElement(By.className("b-products-grid__item-title")).getText();
```
`productNameExpected` is not wrong, but `expectedProductName` is more useful, the same for price.

HaykMelikyan replied:
fixed

HamTorosyan commented on src/test/java/greetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+        // Check name and price
+        Assert.assertEquals(sidebar.findElement(By.tagName("h1")).getText(), productNameExpected,
+                "Name doesn't match. Make sure that the first product in the list hadn't been already added to favorites");
+        Assert.assertEquals(sidebar.findElement(By.className("price-block")).getText(), productPriceExpected,
```
Unreachable line, if Test fails on line 44, to bypass it use `SoftAssert()`

HaykMelikyan replied:
fixed

HamTorosyan commented on src/test/java/greetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+    @Test
+    public void priceAmountTest() throws InterruptedException {
+        // Set driver property, create browser instance, create waiter for that instance
+        System.setProperty("webdriver.chrome.driver", "src/main/resources/chromedriver.exe");
```
You have code repetition on 13, 14 and 59, 60 lines, I would suggest declaring instance of WebDriver outside of all methods (class level), instantiate it inside login method, and use in other methods, as you perform login step in the beginning of other methods

HaykMelikyan replied:
fixed

HamTorosyan reviewed (changes requested):
also use more meaningful branch name according to some conventions

HaykMelikyan committed:
Make corrections
- Change class name
- Change variables' name for expected result
- Perform soft assertion instead of hard
- Avoid code repetition

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+
+        // Get first product's name and price
+        WebElement productItem = driver.findElement(By.cssSelector("div.b-products-grid__item:nth-child(1)"));
+        String ExpectedProductName = productItem.findElement(By.className("b-products-grid__item-title")).getText();
```
Variable name should start by lower-case

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+        productItem.findElement(By.className("b-products-grid__item-action")).click();
+
+        // Go to favorites' page
+        driver.findElement(By.cssSelector("a[class = 'page-header__navigation-item item-user']")).click();
```
Better to have at least half a second sleep between those two steps. It's obviously fast, but test potentially can fail, since the test execution can be even faster

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+
+        // Check name and price
+        SoftAssert assertion = new SoftAssert();
+        assertion.assertEquals(sidebar.findElement(By.tagName("h1")).getText(), ExpectedProductName,
```
`h1` is not a good locator here. It's too fragile and can be changed at any time. You could rely on the class name instead.
Another important thing. Try to avoid nested complicated expressions. They are error-prone, unreadable, and hardly debuggable.

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,115 @@
+
+        // Obtain prices
+        WebElement priceBlock = driver.findElement(By.className("page-detail__price"));
+        String singlePrice = priceBlock.findElement(By.className("price-normal")).getText().substring(2);
```
It's better to use regular expressions. Anyway, this is not an issue, just as a best practice

HaykMelikyan committed:
Add testNG Before-s and After-s
Add Surefire plugin
Make corrections

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,127 @@
+        String actualProductPrice = sidebar.findElement(By.className("price-block")).getText();
+
+        SoftAssert assertion = new SoftAssert();
+        assertion.assertEquals(actualProductName, expectedProductName,
```
Seems, you forgot `assertAll` for Soft Assert. This assertion will not execute

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,127 @@
+        WebElement priceBlock = driver.findElement(By.className("page-detail__price"));
+
+        String singlePrice = priceBlock.findElement(By.className("price-normal")).getText();
+        singlePrice = singlePrice.replaceAll("[^(\\d|\\,)]", "");
```
`,` is not a special character for RegExp pattern. So, you don't need backslash to identify it.

chandranik90 commented on src/test/java/GreetzWebsiteTest.java:
```diff
@@ -0,0 +1,127 @@
+        try {
+            Thread.sleep(2000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
Mark throws in method signature as you did in test methods

HaykMelikyan committed:
Make corrections
